### ISSUE TITLE
baresip: srtp decoding problems with apple clang 16

### ISSUE THREAD

**Issue by njeisecke** [baresip/baresip#3240]
Since building with XCode 16.1 (which brings apple clang 16.0.0) srtp decryption breaks down after some (varying) time, it can be seconds, it can be a few minutes until breakdown. Sending audio continues to work, received audio is just white noise.

This should be reproducible with a default **release** build of a recent baresip version using XCode 16.1 on an Intel Mac (not sure if this also occurs on Apple Silicon).

Workaround is to disable optimization for `srtp/misc.c` in libre:

```
if (APPLE)
  set_source_files_properties(src/srtp/misc.c PROPERTIES COMPILE_FLAGS -O0)
endif()

```

**Comment by sreimers** [comment]
I'll take a wild guess, but can you try:

```diff
diff --git a/src/srtp/misc.c b/src/srtp/misc.c
index 5c43e254..ec1356f6 100644
--- a/src/srtp/misc.c
+++ b/src/srtp/misc.c
@@ -37,7 +37,7 @@ uint64_t srtp_get_index(uint32_t roc, uint16_t s_l, uint16_t seq)
                        v = roc;
        }
 
-       return seq + v*(uint64_t)65536;
+       return seq + v*65536LL;
 }
```

**Comment by njeisecke** [comment]
Thanks. But the problem still occurs.

**Comment by sreimers** [comment]
Ok will try to reproduce next week.

**Comment by njeisecke** [comment]
All testst passed:

```
njeiseck ~/projects/baresip/re/build_tests/test ((v3.18.0))$ ./retest -d ../../test/data -rv
retest: libre version 3.18.0 (x86_64/Darwin)
using async polling method 'kqueue'
using datapath '../../test/data'
regular tests:       test 0 -- test_aac
test 1 -- test_aes
test 2 -- test_aes_gcm
test 3 -- test_aubuf
test 4 -- test_aulength
test 5 -- test_aulevel
test 6 -- test_auposition
test 7 -- test_auresamp
test 8 -- test_async
test 9 -- test_av1
test 10 -- test_dd
test 11 -- test_base64
test 12 -- test_bfcp
test 13 -- test_bfcp_bin
test 14 -- test_bfcp_udp
test 15 -- test_bfcp_tcp
test 16 -- test_conf
test 17 -- test_crc32
test 18 -- test_dns_hdr
test 19 -- test_dns_rr
test 20 -- test_dns_dname
test 21 -- test_dsp
test 22 -- test_dtls
test 23 -- test_dtls_srtp
test 24 -- test_dtmf
test 25 -- test_fir
test 26 -- test_fmt_gmtime
test 27 -- test_fmt_hexdump
test 28 -- test_fmt_human_time
test 29 -- test_fmt_param
test 30 -- test_fmt_pl
test 31 -- test_fmt_pl_alloc_str
test 32 -- test_fmt_pl_float
test 33 -- test_fmt_pl_i32
test 34 -- test_fmt_pl_i64
test 35 -- test_fmt_pl_u32
test 36 -- test_fmt_pl_u64
test 37 -- test_fmt_pl_x3264
test 38 -- test_fmt_print
test 39 -- test_fmt_regex
test 40 -- test_fmt_snprintf
test 41 -- test_fmt_str
test 42 -- test_fmt_str_bool
test 43 -- test_fmt_str_error
test 44 -- test_fmt_str_itoa
test 45 -- test_fmt_str_wchar
test 46 -- test_fmt_timestamp
test 47 -- test_fmt_unicode
test 48 -- test_fmt_unicode_decode
test 49 -- test_g711_alaw
test 50 -- test_g711_ulaw
test 51 -- test_h264
test 52 -- test_h264_sps
test 53 -- test_h264_packet
test 54 -- test_h265
test 55 -- test_h265_packet
test 56 -- test_hash
test 57 -- test_hmac_sha1
test 58 -- test_hmac_sha256
test 59 -- test_http
test 60 -- test_http_loop
test 61 -- test_http_large_body
test 62 -- test_http_conn
test 63 -- test_http_conn_large_body
test 64 -- test_https_loop
test 65 -- test_http_client_set_tls
test 66 -- test_https_large_body
test 67 -- test_https_conn_post_handshake
test 68 -- test_httpauth_chall
test 69 -- test_httpauth_resp
test 70 -- test_httpauth_basic_request
test 71 -- test_httpauth_digest_request
test 72 -- test_httpauth_digest_response
test 73 -- test_httpauth_digest_verification
test 74 -- test_ice_cand
test 75 -- test_ice_loop
test 76 -- test_json
test 77 -- test_json_file
test 78 -- test_json_unicode
test 79 -- test_json_bad
test 80 -- test_json_array
test 81 -- test_list
test 82 -- test_list_flush
test 83 -- test_list_ref
test 84 -- test_list_sort
test 85 -- test_mbuf
test 86 -- test_md5
test 87 -- test_mem
test 88 -- test_mem_reallocarray
test 89 -- test_mem_secure
test 90 -- test_net_if
test 91 -- test_mqueue
test 92 -- test_odict
test 93 -- test_odict_array
test 94 -- test_odict_pl
test 95 -- test_pcp
test 96 -- test_remain
main: thread_init: already added for thread
test 97 -- test_re_assert_se
test 98 -- test_rtmp_play
test 99 -- test_rtmp_publish
test 100 -- test_rtmps_publish
test 101 -- test_rtp
test 102 -- test_rtpext
test 103 -- test_rtcp_encode
test 104 -- test_rtcp_encode_afb
test 105 -- test_rtcp_decode
test 106 -- test_rtcp_decode_badmsg
test 107 -- test_rtcp_packetloss
test 108 -- test_rtcp_twcc
test 109 -- test_rtcp_loop
test 110 -- test_sa_class
test 111 -- test_sa_cmp
test 112 -- test_sa_decode
test 113 -- test_sa_ntop
test 114 -- test_sa_pton
test 115 -- test_sa_pton_linklocal
test 116 -- test_sdp_all
test 117 -- test_sdp_bfcp
test 118 -- test_sdp_parse
test 119 -- test_sdp_oa
test 120 -- test_sdp_extmap
test 121 -- test_sdp_disabled_rejected
test 122 -- test_sha1
test 123 -- test_sip_addr
test 124 -- test_sip_auth
test 125 -- test_sip_drequestf
test 126 -- test_sip_apply
test 127 -- test_sip_hdr
test 128 -- test_sip_param
test 129 -- test_sip_parse
test 130 -- test_sip_via
test 131 -- test_sip_transp_add_client_cert
test 132 -- test_sipevent
test 133 -- test_sipsess
test 134 -- test_sipsess_reject
test 135 -- test_sipsess_blind_transfer
test 136 -- test_sipsess_100rel_caller_require
test 137 -- test_sipsess_100rel_supported
test 138 -- test_sipsess_100rel_answer_not_allowed
test 139 -- test_sipsess_100rel_420
test 140 -- test_sipsess_100rel_421
test 141 -- test_sipsess_update_uac
test 142 -- test_sipsess_update_uas
test 143 -- test_sipsess_update_no_sdp
test 144 -- test_srtp
test 145 -- test_srtcp
test 146 -- test_srtp_gcm
test 147 -- test_srtcp_gcm
test 148 -- test_stun_req
test 149 -- test_stun_resp
test 150 -- test_stun_reqltc
test 151 -- test_stun
test 152 -- test_sys_endian
test 153 -- test_sys_rand
test 154 -- test_sys_fs_isdir
test 155 -- test_sys_fs_isfile
test 156 -- test_sys_fs_fopen
test 157 -- test_sys_getenv
test 158 -- test_tcp
test 159 -- test_telev
test 160 -- test_text2pcap
test 161 -- test_tls
test 162 -- test_tls_ec
test 163 -- test_tls_selfsigned
test 164 -- test_tls_certificate
test 165 -- test_tls_false_cafile_path
test 166 -- test_tls_cli_conn_change_cert
test 167 -- test_tls_session_reuse_tls_v12
test 168 -- test_tls_sni
test 169 -- test_trice_cand
test 170 -- test_trice_candpair
test 171 -- test_trice_checklist
test 172 -- test_trice_loop
test 173 -- test_try_into
test 174 -- test_turn
test 175 -- test_turn_tcp
test 176 -- test_udp
test 177 -- test_unixsock
test 178 -- test_uri
test 179 -- test_uri_encode
test 180 -- test_uri_headers
test 181 -- test_uri_user
test 182 -- test_uri_params_headers
test 183 -- test_uri_escape
test 184 -- test_vid
test 185 -- test_vidconv
test 186 -- test_vidconv_scaling
test 187 -- test_vidconv_pixel_formats
test 188 -- test_websock
test 189 -- test_trace
test 190 -- test_thread
test 191 -- test_dtls_turn
OK
```

Could you give me a pointer or a small patch to enable sequence number printing?

Thanks

**Comment by alfredh** [comment]
here is a small patch to print incoming sequence number:

```diff
diff --git a/src/rtprecv.c b/src/rtprecv.c
index 978348b5..b796614e 100644
--- a/src/rtprecv.c
+++ b/src/rtprecv.c
@@ -380,6 +380,10 @@ void rtprecv_decode(const struct sa *src, const struct rtp_header *hdr,
        bool flush = false;
        int err = 0;
 
+       if (hdr) {
+               re_printf("seq=%u\n", hdr->seq);
+       }
+
        if (!rx)
                return;
 
```


could you also provide debug log and SIP trace ?

which SRTP module does it happen with, `srtp.so` or `dtls_srtp.so` ?

**Comment by njeisecke** [comment]
seq=28172 => initial sequence number
...
seq=32773
seq=32774
seq=32775 audio=64000/64000 (bit/s)
seq=32776

=> looks like the problem starts right here, this can't be a coincidence ;-)

seq=32777
seq=32778

The seemingly random initial sequence number explains the varying duration until the problem occurs.

**Comment by njeisecke** [comment]
Using the "srtp" Module.

**Comment by sreimers** [comment]
Can you retry with XCode 16.2 and `Apple clang version 16.0.0 (clang-1600.0.26.6)`
On my test machine I could not reproduce any problems. Maybe the compiler bug has been fixed.

**Comment by njeisecke** [comment]
Hi, this actually seems to be fixed with clang-1600.0.26.6, looks like clang-1600.0.26.4 was faulty.

Thanks!
